art/project.py

```diff
@@ -192,10 +192,13 @@ def run_all(self, force_rerun=False):
         run_id = get_run_id()
         logger_id = add_logger(EXPERIMENT_LOG_DIR / get_new_log_file_name(run_id))
         try:
-            for step in self.steps:
-                self.metric_calculator.compile(step["skipped_metrics"])
-                step, checks = step["step"], step["checks"]
+            for step_check in self.steps:
+                self.metric_calculator.compile(step_check["skipped_metrics"])
+                step, checks = step_check["step"], step_check["checks"]
                 self.state.current_step = step
+                if isinstance(step, ModelStep):
```

Comment:
I think that project shouldn't be responsible for setting devices. I think perfect place to put it is inside `ModelStep.__call__` . We have access to both metric calculator and trainer
![image](https://github.com/SebChw/Actually-Robust-Training/assets/66476124/8d04cf2f-88c1-4314-92bf-e33654b9388f)

Reply:
Oh great, I didn't see the metric calculator there